### PR TITLE
set public env in `<script>`, implementing preloading on safari

### LINKED ISSUE #8946: Omit `env` from server renders if `$env/dynamic/public` isn't used
### Describe the problem

Every public environment variable is made available to client-side code via `$env/dynamic/public` in the init block:

```html
<script type="module" data-sveltekit-hydrate="gb5305">
  import { start } from "../_app/immutable/start-02c2bf02.js";

  start({
    assets: "",
    env: {
      PUBLIC_FOO: 'a potentially very long string',
      PUBLIC_BAR: 'another potentially very long string',
      PUBLIC_BAZ: 'yet another potentially very long string. i could do this all day'
    },
    // ...
  });
</script>
```

But many apps don't actually use `$env/dynamic/public`, they use `$env/static/public` instead (which in general is preferred). 

### Describe the proposed solution

Detect cases where `$env/dynamic/public` is not imported by client-side code, and omit public env vars from the init block.

### Alternatives considered

_No response_

### Importance

nice to have

### Additional Information

_No response_

### LINKED ISSUE #8997: Eagerly import modules for the current route
### Describe the problem

While working on #8957 I was reminded that `modulepreload` links don't work outside Chromium browsers. (Vite provides a polyfill but we don't use it, but that's probably for the best: as far as I've been able to tell it doesn't actually work, but rather causes duplicate requests.)

The result is that imports create a waterfall — module `a` isn't requested until its importer `b` is available, which isn't requested until its importer `c` is available. In Chrome, `a`, `b` and `c` can all be requested concurrently, resulting in faster application startup.

#8957 exacerbates this problem by separating the app code into a separate chunk that doesn't get imported until the framework code is evaluated.

### Describe the proposed solution

I think the solution is actually quite straightforward — instead of trying to polyfill `modulepreload` with `fetch`, we just import the modules in question.

In the general case you can't do that because the semantics are different — `modulepreload` doesn't cause modules to be evaluated, whereas `import` does. But in our case that's totally fine, because we know the modules were going to be evaluated _anyway_.

(We should make sure that this only includes modules that are statically imported by the current route — we don't want to eagerly import a module that is ordinarily only imported when a button is clicked, for example.)

The one big problem is the issue that #8957 set out to solve in the first place: if we eagerly import app code, we need to make sure that `$env/dynamic/public` is populated before it runs. There's a couple of ways this could work:

## 1. Pollute `window`

The easiest way is probably to inject a script at the top of the `<head>`...

```html
<script>sk_env_xyz123={PUBLIC_FOO:'bar'}</script>
```

...and update `$env/dynamic/public` accordingly:

```js
export const env = sk_env_xyz123;
```

Here, `xyz123` is a hash of the version name, preventing conflicts even if the app uses `embedded: true`.

## 2. Indirection

This would be ideal...

```js
// $env/dynamic/private
export { env } from '<assets>/_app/env.js';
```

...except that it wouldn't work with relative asset paths, and might be hard to get to work with Vite. (It creates a new module that needs to be requested, though it also solves #8946 quite neatly.)

### Alternatives considered

_No response_

### Importance

nice to have

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by Rich-Harris**
* closes #8946 
* closes #8997

This changes how `$env/dynamic/public` is populated in such a way that we can eagerly import the modules required for the current route.

In the case where `paths.assets` is not specified, we can generate `_app/env.js` dynamically (except in the `adapter-static` case — will need a `generatePublicEnv()` method for that). By rejiggering the Rollup output such that all modules inside `_app/immutable` are at the same depth, we can use a combination of Rollup's `external` and `output.paths` config to turn any references to `$env/dynamic/public` into relative imports.

Where `paths.assets` _is_ specified, we don't have this luxury. In those cases we need to populate `$env/dynamic/public` the same way we currently do, by setting it from a `<script>` on the page. In order to ensure it's ready before any eagerly imported modules evaluate, we must assign it to a global variable at the top of the `<head>`.

* [x] create the eager import block
* [x] add `generatePublicEnv()` method

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [ ] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [ ] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### CODE REVIEW

**Comment by Rich-Harris** [comment]
Gah, it's less simple than I thought. We actually _can't_ evaluate stuff eagerly because `client` needs to be instantiated before any user code runs. (We also need to call `set_assets` and `set_version`, though those could be solved more easily.)

We _could_ instantiate `client` immediately, but then we need to handle edge cases where someone calls `goto` before the app has started...

Dammit why can't browsers just implement `modulepreload`

**Comment by Rich-Harris** [comment]
I have made a discovery!

In Safari, `<link rel="preload" as="script" href="x.js" crossorigin="anonymous">` behaves very similarly to `modulepreload`. It won't cause dependencies to also be preloaded, but that's okay because we already flatten the graph.

Which means all we need to do to effectively polyfill `modulepreload` is this:

```html
<script>
  const { relList } = document.createElement('link');
  if (!relList?.supports('modulepreload')) {
    for (const link of document.querySelectorAll('link[rel="modulepreload"]')) {
      const preload = document.createElement('link');
      preload.href = link.href;
      preload.rel = 'preload';
      preload.as = 'script';
      preload.crossOrigin = 'anonymous';
      document.head.appendChild(preload);
    }
  }
</script>
```

(In Firefox, this will cause the files to be requested twice, because reasons. But tough shit Firefox! This is what you get for not implementing `modulepreload`. Unfortunately I'm one of your last dozen or so users, so we have to prioritise the needs of people using Safari on iOS.)

In reality we probably wouldn't `document.querySelectorAll('link[rel="modulepreload"]')` because `<link>` is only rendered when prerendering (otherwise it is expressed as an HTTP `Link` header), so we probably need to do this instead:

```html
<script>
  const { relList } = document.createElement('link');
  if (!relList?.supports('modulepreload')) {
    for (const src of ${s(modules)}) {
      const link = document.createElement('link');
      link.rel = 'preload';
      link.as = 'script';
      link.href = src;
      link.crossOrigin = 'anonymous';
      document.head.appendChild(link);
    }
  }
</script>
```

With this approach, we don't need to eagerly evaluate anything.

**Comment by Rich-Harris** [comment]
9b71ed931fb804465320d9d309bc751aa2f3a669 implements the idea above. It results in a nice speedup on Safari:

## Before

Screenshot 2023-02-15 at 8 45 37 PM

(The stragglers in the lower right are lazily imported by design and can be safely ignored.)

## After

Screenshot 2023-02-15 at 8 46 13 PM

---

As mentioned, it does cause double requests in Firefox, but the needs of the many outweigh the needs of the few. I think this is ready to merge into #8957.

File: packages/kit/src/core/sync/write_server.js [hunk 1]
```diff
@@ -30,6 +31,7 @@ import { set_assets, set_building, set_private_env, set_public_env, set_version
 set_version(${s(config.kit.version.name)});
 
 export const options = {
+	app_template_contains_nonce: ${template.includes('%sveltekit.nonce%')},
```
Comment (Rich-Harris) [inline packages/kit/src/core/sync/write_server.js:34]: unrelated drive-by fix

**Comment by jridgewell** [comment]
If you use `.mjs` extension, you actually can get the automatic module parsing in Chrome using `rel=prelaod as=script`: [demo](https://modulepreload-demo-8hdvvz44g-jridgewell.vercel.app/), [code](https://github.com/jridgewell/modulepreload-demo), [chrome code](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/bindings/core/v8/script_streamer.cc;l=799-808;drc=dff14563c5f4f1c4a0ac2d4360165239221a9321)

```html
<link rel="preload" as="script" crossorigin="anonymous" href="./a.mjs" />
<link rel="preload" as="script" crossorigin="anonymous" href="./b.mjs" />
<link rel="preload" as="script" crossorigin="anonymous" href="./c.mjs" />

<!-- snip… -->
<script type="module" crossorigin="anonymous" src="./a.mjs"></script>
```

This gives Chrome optimal performance, Safari gets the default preload behavior with single fetches (without the `relList` checks), and FireFox is still broken with double fetches.

File: packages/kit/src/exports/vite/index.js [hunk 2]
```diff
@@ -351,6 +373,13 @@ function kit({ svelte_config }) {
 						vite_config_env.command === 'serve' ? env.private : undefined
 					);
 				case '\0$env/dynamic/public':
+					// populate `$env/dynamic/public` from `window` in the case where
+					// modules are externally hosted
+					if (kit.paths.assets && is_build && !options?.ssr) {
+						// TODO use a hash of the version name to avoid conflicts
+						return `export const env = window.__sveltekit_${version_hash}.env;`;
```
Comment (dummdidumm) [inline packages/kit/src/exports/vite/index.js:380]: In #8901 we also write to the global window. We should put these things below a common top level key. I went with `$__sveltekit__`. We can put the hash at the top level (like you do) or one level below.

Btw the TODO (and one below where you set the env) is probably outdated.

**Review by dummdidumm — COMMENTED** [review]
I'm wondering in general how we can make things like these more understandable. The logic is spread out into several files. For example I didn't understand at first why the script is only added within the `if (assets)` condition in `render.js`. Ideas:
- conservative: add a comment above each of these linking to this PR which contains an explanation, or linking to a comment that explains things more thoroughly
- experimental: move all env related stuff into one file, have some explanation there, and everyone that uses stuff from it imports from that file. That way it's colocated and hopefully better understandable. On the flip side, things are no longer where they are needed (`render.js` for example) and often only imported from one location, false sense of reusability. I guess we would need to see how it "feels" to do this

File: packages/kit/src/exports/vite/index.js (hunk 2 above)
Comment (Rich-Harris) [reply packages/kit/src/exports/vite/index.js:380]: Good catch. I think top-level is probably easiest since we then don't need to check to see if `$__sveltekit__` already exists before creating it/adding the key

Comment (dummdidumm) [reply packages/kit/src/exports/vite/index.js:380]: Is there a potential use case in the future (programmatic navigation from outside for example) that warrants a stable key? What conflicts do you had in mind when creating this? Multiple embedded SvelteKit apps?

File: packages/kit/src/exports/vite/index.js [hunk 3]
```diff
@@ -491,9 +518,9 @@ export function set_assets(path) {
 							input,
 							output: {
 								format: 'esm',
-								entryFileNames: ssr ? '[name].js' : `${prefix}/[name]-[hash].js`,
-								chunkFileNames: ssr ? 'chunks/[name].js' : `${prefix}/chunks/[name]-[hash].js`,
-								assetFileNames: `${prefix}/assets/[name]-[hash][extname]`,
+								entryFileNames: ssr ? '[name].js' : `${prefix}/[name].[hash].js`,
+								chunkFileNames: ssr ? 'chunks/[name].js' : `${prefix}/chunks/[name].[hash].js`,
+								assetFileNames: `${prefix}/assets/[name].[hash][extname]`,
```
[the remark this answers is not in the document]
Comment (Rich-Harris) [reply packages/kit/src/exports/vite/index.js:523]: It's a minor thing, but between these...

```
sverdle-how-to-play-page.js-13bb5e39.js
sverdle-how-to-play-page.js.13bb5e39.js
```

...the second makes more sense to me, because you have four sections of the filename:

```
[route].[ext].[hash].js
```

When they're joined by a hyphen, the extension and the hash become one section which doesn't really make sense

**Comment by Rich-Harris** [comment]
> If you use `.mjs` extension, you actually can get the automatic module parsing in Chrome using `rel=prelaod as=script`: [demo](https://modulepreload-demo-8hdvvz44g-jridgewell.vercel.app/), [code](https://github.com/jridgewell/modulepreload-demo), [chrome code](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/bindings/core/v8/script_streamer.cc;l=799-808;drc=dff14563c5f4f1c4a0ac2d4360165239221a9321)

Oh wow! What a fun easter egg. I had hoped perhaps the same thing would work with `Link` headers (since that's what we use unless we're prerendering content). Turns out we get the ideal behaviour with Safari, but Chrome behaves _very_ strangely:

image

So unfortunately we can't just update both `Link` and `<link>` to use `preload` instead of `modulepreload` — we need to continue using `Link: </whatever.js> ref="modulepreload"`. But [according to my tests](https://modulepreload-demo-git-modulepreload-header-and-p-bedc90-svelte.vercel.app/) we can use that alongside `<link rel="preload">` without any double-loading penalty — this gets us the optimal behaviour in Chrome and Safari.

File: packages/kit/src/runtime/server/page/render.js [hunk 4]
```diff
@@ -318,24 +313,38 @@ export async function render_response({
...
+			import { start } from ${s(prefixed(client.start.file))};
 
 			start({
 				${opts.join(',\n\t\t\t\t')}
 			});
 		`;
 
-		for (const dep of modulepreloads) {
-			const path = prefixed(dep);
+		const filtered_modulepreloads = Array.from(modulepreloads, dep => prefixed(dep)).filter(path => resolve_opts.preload({ type: 'js', path }));
```
[the remark this answers is not in the document]
Comment (Rich-Harris) [reply packages/kit/src/runtime/server/page/render.js]: yeah, i changed it to `included_modulepreloads`

File: packages/kit/src/exports/vite/index.js (hunk 2 above)
Comment (Rich-Harris) [reply packages/kit/src/exports/vite/index.js:380]: Yeah, I was thinking about embeds mainly

Comment (Rich-Harris) [reply packages/kit/src/exports/vite/index.js:379]: removed the TODO comments

**Comment by Rich-Harris** [comment]
> I'm wondering in general how we can make things like these more understandable

it's a good point. i added some comments to hopefully make things clearer (and `git blame` can always point the way back here). in general i think colocation is a net negative if it results in greater indirection, which it would here

**Comment by dummdidumm** [comment]
Mhm shit I just realized that adding the new adapter API means we need a major version bump of adapter static, but I think it isn't backwards compatible such that the current adapter static version can continue to work without it. What now?
